### PR TITLE
Patch BoringSSL files for C tests

### PR DESCRIPTION
**Pull request by JasonGross**
See https://github.com/mit-plv/fiat-crypto/pull/1684#issuecomment-1774148899

### CODE REVIEW

**Comment by JasonGross** [comment]
@andres-erbsen I'm going to set this to squash and merge so we can (relatively) quickly unblock the other failing PRs, but feel free to revert and implement some other preferred alternative.

File: etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
++#if !defined(OPENSSL_NO_ASM) && defined(__GNUC__) && defined(__x86_64__)
++  if (CRYPTO_is_BMI1_capable() && CRYPTO_is_BMI2_capable() &&
++    CRYPTO_is_ADX_capable()) {
++      fiat_p256_adx_sqr(out1, arg1);
++      return;
++  }
++#endif
+   internal_fiat_p256_square((uintptr_t)out1, (uintptr_t)arg1);
+ }
+
```
Comment (JasonGross) [inline etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch:42]: @andres-erbsen Is this wrong for bedrock2?
```
FAILED: CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o 
  /usr/bin/c++ -DBORINGSSL_DISPATCH_TEST -DBORINGSSL_HAVE_LIBUNWIND -DBORINGSSL_IMPLEMENTATION -I/home/runner/work/fiat-crypto/fiat-crypto/boringssl/third_party/googletest/googlemock/include -I/home/runner/work/fiat-crypto/fiat-crypto/boringssl/third_party/googletest/googletest/include -I/home/runner/work/fiat-crypto/fiat-crypto/boringssl/include -Wno-error=unused-function -Wno-error=unused-but-set-variable -Werror -Wformat=2 -Wsign-compare -Wwrite-strings -Wvla -Wshadow -Wtype-limits -ggdb -Wall -fvisibility=hidden -fno-common -Wno-free-nonheap-object -Wimplicit-fallthrough -Wformat-signedness -Wmissing-declarations -fno-exceptions -fno-rtti -std=gnu++14 -MD -MT CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o -MF CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o.d -o CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o -c /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc
  In file included from /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/../../../third_party/fiat/p256_64.h:28,
                   from /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc:22:
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/../../../third_party/fiat/p256_64.h: In function ‘void _br2_preconditions()’:
  Error: /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/../../../third_party/fiat/p256_64.h:33:75: error: taking address of rvalue [-fpermissive]
     33 |   assert(((void)"little-endian", 1 == *(unsigned char *)&(const uintptr_t){1}));
        |                                                                           ^~~
  Error: /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/../../../third_party/fiat/p256_64.h:34:74: error: taking address of rvalue [-fpermissive]
     34 |   assert(((void)"little-endian", 1 == *(unsigned char *)&(const intptr_t){1}));
        |                                                                          ^~~
```

**Comment by andres-erbsen** [comment]
I don't understand what you're trying to do here. IIUC the point of this integration test is to exercise the bedrock2-generated C code, so we probably shouldn't prefer assembly implementations?

File: etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch (hunk 1 above)
Comment (andres-erbsen) [reply etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch:42]: This error looks like bedrock2-generated non-portable C failing to compile with strict flags from BoringSSL. I don't think it's related to the patch you are commenting on.

Comment (andres-erbsen) [reply etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch:42]: https://github.com/mit-plv/bedrock2/pull/382

Comment (JasonGross) [reply etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch:42]: I guess instead we should be passing `-DOPENSSL_NO_ASM` to the build process, and then we don't have to patch the files at all to fix
```
FAILED: CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o 
  /usr/bin/c++ -DBORINGSSL_DISPATCH_TEST -DBORINGSSL_HAVE_LIBUNWIND -DBORINGSSL_IMPLEMENTATION -I/home/runner/work/fiat-crypto/fiat-crypto/boringssl/third_party/googletest/googlemock/include -I/home/runner/work/fiat-crypto/fiat-crypto/boringssl/third_party/googletest/googletest/include -I/home/runner/work/fiat-crypto/fiat-crypto/boringssl/include -Wno-error=unused-function -Werror -Wformat=2 -Wsign-compare -Wwrite-strings -Wvla -Wshadow -Wtype-limits -ggdb -Wall -fvisibility=hidden -fno-common -Wno-free-nonheap-object -Wimplicit-fallthrough -Wformat-signedness -Wmissing-declarations -fno-exceptions -fno-rtti -std=gnu++14 -MD -MT CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o -MF CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o.d -o CMakeFiles/crypto_test.dir/crypto/fipsmodule/ec/p256_test.cc.o -c /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc
  In file included from /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc:17:
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc: In member function ‘virtual void P256Test_AdxMulABI_Test::TestBody()’:
  Error: /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc:30:15: error: ‘fiat_p256_adx_mul’ was not declared in this scope; did you mean ‘fiat_p256_mul’?
     30 |     CHECK_ABI(fiat_p256_adx_mul, out, in1, in2);
        |               ^~~~~~~~~~~~~~~~~
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/../../test/abi_test.h:401:34: note: in definition of macro ‘CHECK_ABI_SEH’
    401 |                                  __VA_ARGS__)
        |                                  ^~~~~~~~~~~
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc:30:5: note: in expansion of macro ‘CHECK_ABI’
     30 |     CHECK_ABI(fiat_p256_adx_mul, out, in1, in2);
        |     ^~~~~~~~~
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc: In member function ‘virtual void P256Test_AdxSquareABI_Test::TestBody()’:
  Error: /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc:42:15: error: ‘fiat_p256_adx_sqr’ was not declared in this scope; did you mean ‘fiat_p256_add’?
     42 |     CHECK_ABI(fiat_p256_adx_sqr, out, in);
        |               ^~~~~~~~~~~~~~~~~
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/../../test/abi_test.h:401:34: note: in definition of macro ‘CHECK_ABI_SEH’
    401 |                                  __VA_ARGS__)
        |                                  ^~~~~~~~~~~
  /home/runner/work/fiat-crypto/fiat-crypto/boringssl/crypto/fipsmodule/ec/p256_test.cc:42:5: note: in expansion of macro ‘CHECK_ABI’
     42 |     CHECK_ABI(fiat_p256_adx_sqr, out, in);
        |     ^~~~~~~~~
  [284/754] Building CXX object CMakeFiles/crypto_test.dir/crypto/fipsmodule/ecdsa/ecdsa_test.cc.o
  [285/754] Building CXX object CMakeFiles/crypto_test.dir/crypto/fipsmodule/hkdf/hkdf_test.cc.o
  ninja: build stopped: subcommand failed.
```
?

How do we edit
```
cmake -GNinja .. -DCMAKE_CXX_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" -DCMAKE_C_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" || exit $?
ninja || exit $?
```
to do this?  Do we want to pass `-DOPENSSL_NO_ASM` to `cmake`, or do we want to stick it in `CMAKE_CXX_FLAGS` or `CMAKE_C_FLAGS`...?

Comment (andres-erbsen) [reply etc/ci/boringssl-bedrock2-patches/2023-10-05-p256-adx.patch:42]: `env CC=clang CXX=clang++ cmake -GNinja -B build-noasm -DOPENSSL_NO_ASM=1 -DCMAKE_BUILD_TYPE=Release` seems to work for me

File: etc/ci/test-fiat-c-boringssl.sh [hunk 2]
```diff
@@ -40,7 +40,7 @@ echo "::group::Building BoringSSL"
     set -ex
     mkdir build
     cd build
-    cmake -GNinja .. -DCMAKE_CXX_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" -DCMAKE_C_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" || exit $?
+    cmake -GNinja .. -DOPENSSL_NO_ASM=1 -DCMAKE_CXX_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" -DCMAKE_C_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" || exit $?
```
Comment (JasonGross) [inline etc/ci/test-fiat-c-boringssl.sh:43]: @andres-erbsen you suggested
> `env CC=clang CXX=clang++ cmake -GNinja -B build-noasm -DOPENSSL_NO_ASM=1 -DCMAKE_BUILD_TYPE=Release` seems to work for me

What's the relationship between `..` and `-B build-noasm` (if any)?  What's the importance of `-B build-noasm`?

And what are your thoughts on `-DCMAKE_BUILD_TYPE=Release` vs `-DCMAKE_CXX_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}" -DCMAKE_C_FLAGS="-Wno-error=unused-function ${EXTRA_CFLAGS}"`?  (`EXTRA_CFLAGS` gets set to `-Wno-error=unused-but-set-variable` by CI)

**Comment by JasonGross** [comment]
Apparently build time for C tests jumped from ~13m on the last successful run before this change to ~30m after this change, did BoringSSL get much slower to build around this time? @davidben @andres-erbsen

**Comment by davidben** [comment]
Can't think of anything. We have been adjusting the build, but I don't see anything obviously relevant around October. I dunno, if it's a huge problem, maybe try building it at different versions and see what changed?

**Pull request merged**